tests/models/glm/test_modeling_glm.py

```diff
@@ -0,0 +1,955 @@
+                self.skipTest(f"{model_class.__name__} with sliding window attention is not supported by this test")
+            model = model_class(config).to(device=torch_device, dtype=torch.float32)
+
+            (
+                input_ids,
+                position_ids,
+                input_ids_shared_prefix,
+                mask_shared_prefix,
+                position_ids_shared_prefix,
+            ) = self._get_custom_4d_mask_test_data()
+
+            logits = model.forward(input_ids, position_ids=position_ids).logits
+            # logits.shape == torch.Size([3, 4, ...])
+
+            logits_shared_prefix = model(
+                input_ids_shared_prefix,
+                attention_mask=mask_shared_prefix,
+                position_ids=position_ids_shared_prefix,
+            )[0]
+            # logits_shared_prefix.shape == torch.Size([1, 6, ...])
+
+            out_last_tokens = logits[:, -1, :]  # last tokens in each batch line
+            out_shared_prefix_last_tokens = logits_shared_prefix[0, -3:, :]  # last three tokens
+
+            # comparing softmax-normalized logits:
+            normalized_0 = torch.nn.functional.softmax(out_last_tokens)
+            normalized_1 = torch.nn.functional.softmax(out_shared_prefix_last_tokens)
+            print(torch.abs(normalized_0 - normalized_1).max())
+
+            torch.testing.assert_close(normalized_0, normalized_1, rtol=1e-3, atol=1e-3)
+
+    @require_flash_attn
+    @require_torch_gpu
+    @pytest.mark.flash_attn_test
+    @slow
+    def test_flash_attn_2_generate_padding_right(self):
+        """Overwrite the common test as the test is flaky on tiny models."""
+        model = GlmForCausalLM.from_pretrained(
+            "THUDM/glm-4-9b",
+            device_map={"": 0},
+            torch_dtype=torch.bfloat16,
+            revision="refs/pr/15",
+        )
+
+        tokenizer = AutoTokenizer.from_pretrained("THUDM/glm-4-9b", revision="refs/pr/15")
+        tokenizer.padding_side = "right"
+
+        texts = ["hi", "Hello this is a very long sentence"]
+        inputs = tokenizer(texts, return_tensors="pt", padding=True).to(0)
+
+        output_native = model.generate(**inputs, max_new_tokens=15, do_sample=False)
+        output_native = tokenizer.batch_decode(output_native)
+
+        model = GlmForCausalLM.from_pretrained(
+            "THUDM/glm-4-9b",
+            device_map={"": 0},
+            attn_implementation="flash_attention_2",
+            torch_dtype=torch.bfloat16,
+            revision="refs/pr/15",
+        )
+
+        output_fa_2 = model.generate(**inputs, max_new_tokens=15, do_sample=False)
+        output_fa_2 = tokenizer.batch_decode(output_fa_2)
+
+        self.assertListEqual(output_native, output_fa_2)
+
+    @parameterized.expand([("float16",), ("bfloat16",), ("float32",)])
+    @require_torch_sdpa
+    @slow
+    @is_flaky
+    def test_eager_matches_sdpa_inference(self, torch_dtype: str):
```

Comment:
why do we have to overwrite this one?

Reply:
Unfortunately, based on the random inputs there may be sometimes when one of the tests fail - this I overwrote it to add the `flaky` decorator which allows the test to consistently pass